dags/process_md_1c_to_pg/dag.py

```diff
@@ -0,0 +1,94 @@
+import json
+import logging
+import os
+import uuid
+
+from airflow import DAG
+from airflow.exceptions import AirflowException
+from airflow.sdk import task, teardown, Variable
+from datetime import datetime
+
+from constants import TZ_MSK
+from process_md_1c_to_pg.libs.transform import transform_md_data
+from process_md_1c_to_pg.libs.upload import PgMdHook
+
+with DAG(
+    dag_id="process_md_1c_to_pg",
+    start_date=datetime(2025, 5, 1, tzinfo=TZ_MSK),
+    schedule='30 8 * * 1-5',
+    catchup=False,
+    tags=['1c', 'md', 'postgresql']
+) as dag:
+    
+    def get_local_tmp_dir_path():
+        return os.path.join(Variable.get('tmp_dir_path'), 'md_1c')
+
+    @task
+    def download_task() -> str:
+        from airflow.providers.sftp.hooks.sftp import SFTPHook
+
+        local_dp = get_local_tmp_dir_path()
+        os.makedirs(local_dp, exist_ok=True)
+        
+        sftp_hook = SFTPHook("sftp_1c") 
+        
+        files_to_download = {
+            "products": Variable.get("md_1c_products_sftp_path"),
+            "price_list": Variable.get("md_1c_pricelist_sftp_path")
+        }
+        
+        local_filepaths = {}
+        failed_keys = []
+
+        for key, remote_fp in files_to_download.items():
+            if not remote_fp:
+                logging.info("SFTP path for %s is not configured. Skipping.", key)
+                continue
+            local_fp = os.path.join(local_dp, f"{uuid.uuid4().hex}_{os.path.basename(remote_fp)}")
+            try:
+                sftp_hook.retrieve_file(remote_fp, local_fp)
+                local_filepaths[key] = local_fp
+                logging.info("Downloaded %s to %s", remote_fp, local_fp)
+            except FileNotFoundError:
+                failed_keys.append(key)
+                logging.error("Failed to download required file on SFTP: %s", remote_fp)
+        
+        # Fail the DAG if at least one expected file failed to download
+        configured_expected_keys = [k for k, path in files_to_download.items() if path]
+        if len(local_filepaths) < len(configured_expected_keys):
+            missing = sorted(set(configured_expected_keys) - set(local_filepaths.keys())) or failed_keys
+            raise AirflowException(f"Not all files were downloaded. Missing: {missing}")
```

Comment:
Какой практический смысл в вычислении разницы между сетами? Почему нельзя просто использовать проверку на failed_keys?
```suggestion
        if failed_keys:
            raise AirflowException(f"Not all files were downloaded. Missing: {failed_keys}")
```

Reply:
Разница сетов нужна, чтобы зафиксировать факт, что ожидаемый файл отсутствует в результате x, вне зависимости от причины. failed_keys будто покрывает только конкретный путь (например file not found). Если ключ не попал в local_filepaths по другой причине (другой тип исключения, ранний выход, забыли записать в словарь), простая проверка failed_keys это не поймает. Дополнительный if failed_keys: после текущего raise всё равно не выполнится или задублирует сообщение. Может упростить и оставить единую проверку по разнице, опционально добавив failed_keys в детали сообщения? :

```
expected = {k for k, p in files_to_download.items() if p}
actual = set(local_filepaths.keys())
missing = sorted(expected - actual)

if missing or failed_keys:
    details = {'missing': missing}
    if failed_keys:
        details['failed_keys'] = failed_keys
    raise AirflowException(f"Not all files were downloaded. Details: {details}")
```